bssl-compat/source/ssl.c

```diff
@@ -85,79 +87,76 @@ int SSL_get_error(const SSL *ssl, int ret_code) {
+  if (ossl_mode & ossl_SSL_MODE_SEND_FALLBACK_SCSV)
+    boringssl_mode |= SSL_MODE_SEND_FALLBACK_SCSV;
 
-	 if (mode & openssl_SSL_MODE_ASYNC)
-		 boringssl_mode |= SSL_MODE_ASYNC;
+  /* The following flags are in OpenSSL but not in BoringSSL */
+  if (ossl_mode & ossl_SSL_MODE_ASYNC)
+    bssl_compat_fatal("SSL_MODE_ASYNC has no BoringSSL equivalent");
+  if (ossl_mode & ossl_SSL_MODE_DTLS_SCTP_LABEL_LENGTH_BUG)
+    bssl_compat_fatal("SSL_MODE_DTLS_SCTP_LABEL_LENGTH_BUG has no BoringSSL equivalent");
+  if (ossl_mode & ossl_SSL_MODE_AUTO_RETRY)
+    bssl_compat_fatal("SSL_MODE_AUTO_RETRY has no BoringSSL equivalent");
+  if (ossl_mode & ossl_SSL_MODE_RELEASE_BUFFERS)
+    bssl_compat_fatal("SSL_MODE_RELEASE_BUFFERS has no BoringSSL equivalent");
+  if (ossl_mode & ossl_SSL_MODE_SEND_CLIENTHELLO_TIME)
+    bssl_compat_fatal("SSL_MODE_SEND_CLIENTHELLO_TIME has no BoringSSL equivalent");
+  if (ossl_mode & ossl_SSL_MODE_SEND_SERVERHELLO_TIME)
+    bssl_compat_fatal("SSL_MODE_SEND_SERVERHELLO_TIME has no BoringSSL equivalent");
 
 	 return boringssl_mode;
 }
 
 /* #define SSL_set_mode(ssl,op) SSL_ctrl((ssl),SSL_CTRL_MODE,(op),NULL) */
 uint32_t SSL_set_mode(SSL *ssl, uint32_t mode) {
 	uint32_t openssl_mode = 0;
-	uint32_t boringssl_mode = 0;
+	uint32_t boringssl_mode = SSL_get_mode(ssl);
 
 	if (mode & SSL_MODE_ENABLE_PARTIAL_WRITE) {
-		openssl_mode |= openssl_SSL_MODE_ENABLE_PARTIAL_WRITE;
+		openssl_mode |= ossl_SSL_MODE_ENABLE_PARTIAL_WRITE;
 		boringssl_mode |= SSL_MODE_ENABLE_PARTIAL_WRITE;
 	}
 
 	if (mode & SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER) {
-		openssl_mode |= openssl_SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER;
+		openssl_mode |= ossl_SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER;
 		boringssl_mode |= SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER;
 	}
 
 	if (mode & SSL_MODE_NO_AUTO_CHAIN) {
-		openssl_mode |= openssl_SSL_MODE_NO_AUTO_CHAIN;
+		openssl_mode |= ossl_SSL_MODE_NO_AUTO_CHAIN;
 		boringssl_mode |= SSL_MODE_NO_AUTO_CHAIN;
 	}
 
 	if (mode & SSL_MODE_SEND_FALLBACK_SCSV) {
-		openssl_mode |= openssl_SSL_MODE_SEND_FALLBACK_SCSV;
+		openssl_mode |= ossl_SSL_MODE_SEND_FALLBACK_SCSV;
 		boringssl_mode |= SSL_MODE_SEND_FALLBACK_SCSV;
 	}
 
-	if (mode & SSL_MODE_ASYNC) {
-		openssl_mode |= openssl_SSL_MODE_ASYNC;
-		boringssl_mode |= SSL_MODE_ASYNC;
-	}
+  if(mode & SSL_MODE_ENABLE_FALSE_START)
+    bssl_compat_fatal("SSL_MODE_ENABLE_FALSE_START (or SSL_MODE_HANDSHAKE_CUTTHROUGH) is not supported by OpenSSL");
+  if(mode & SSL_MODE_CBC_RECORD_SPLITTING)
+    bssl_compat_fatal("SSL_MODE_CBC_RECORD_SPLITTING is not supported by OpenSSL");
+  if(mode & SSL_MODE_NO_SESSION_CREATION)
+    bssl_compat_fatal("SSL_MODE_NO_SESSION_CREATION is not supported by OpenSSL");
 
-	openssl.SSL_ctrl(ssl, SSL_CTRL_MODE, openssl_mode, NULL);
+	ossl_SSL_ctrl(b2o(ssl), ossl_SSL_CTRL_MODE, openssl_mode, NULL);
 
 	return boringssl_mode;
 }
 
-/* SSL_get_all_async_fds exists only in OpenSSL */
-int SSL_get_all_async_fds(SSL *s, OSSL_ASYNC_FD *fds, size_t *numfds) {
```

Comment:
This function is needed by the Envoy handshake extension in openssl_handshaker/. It is indeed not supported by BoringSSL, but how should this be handled in the bssl bridge library code?

Reply:
In this PR, there is no `SSL_get_all_async_fds()` function declared in the public headers, because those public headers are copied directly from BoringSSL.
To solve this we could either:
1. Have the handshaker code call the `ossl_SSL_get_all_async_fds(b20(ssl), ...)` i.e. call the OpenSSL function directly.
2. Add a "BoringSSL" version of the function by patching a declaration into `openssl/ssl.h` and reinstating this implementation in `ssl.c`.

Reply:
On one hand SSL_xxx() functions use the SSL * in their argument list, which is also true for SSL_get_all_async_fds(). It would look inconsistent if the ossl_SSL_xxx() and SSL_yyy() functions were to share the same SSL * argument?
On the other hand calling ossl_SSL_xxx() raises the flag that this code is outside the BoringSSL boundary where other things not expected by BoringSSL may happen.

Reply:
Yes, that would be inconsistent. The `SSL` and `ossl_SSL` types are distinct, and the compiler will complain if a `SSL*` is passed into an `ossl_SSL_xxx()` function, or if an `ossl_SSL*` is passed to a `SSL_xxx()` function.

I have added `<ext/openssl/ssl.h>` which declares `ext_SSL_get_all_async_fds()` and updated `openssl_handshaker.c`. We can follow the same pattern for any other similar cases